### ISSUE TITLE
bug: app renders unusably slowly

### ISSUE THREAD

**Issue by shrouxm** [techmatters/terraso-mobile-client#120]
## Description
This is seen most easily by trying to switch between different screens on the bottom tabs, but also applies when leaving form screens (e.g. Create Site), navigating to nested screens (e.g. navigating through nested data input screens), or causing screen updates (e.g. forms with updating helper text). The behavior is intermittent but has been getting overall worse over time.

**Comment by ltseng** [comment]
Still replicable on version 40 on iOS and Android

**Comment by shrouxm** [comment]
@ltseng could you provide a bit more info? i haven't really been seeing this issue and first reported it when the app was very new, so i'm not sure what form this bug takes anymore. specific questions:
- are there particular screen transitions that are slow, or could it be any screen transition?
- how slow is the screen transition? (less than one second but noticeable lag? more than one second? inconsistent?)
- how often does it happen? (every time? some of the time? rarely?)

**Comment by ltseng** [comment]
I specifically see this on iPhone when switching between the Home and Projects tab. I'm not sure if it's because the touch target is somehow smaller on the iPhone, but when there is a delay, it's consistently less than a second, and sometimes the screen fails to switch at all.

**Comment by shrouxm** [comment]
ok thank you for the extra info! from watching the video, i'm guessing it's because of missing the touch target

we could try to resolve by making the touch targets as big as possible (i.e. extend all the way to the top and bottom of the bar, and halfway to the next button in the bar horizontally). that could potentially introduce a different frustrating interaction where users accidentally change screens when trying to tap elements near the bottom bar, so there might be a trade-off there

it seems like it happens pretty infrequently at this point and the work to do the fix is straightforward but the results might be finnicky so i will update the issue to reflect that

**Comment by ltseng** [comment]
Yeah, I wouldn't consider the bug that I encountered to be a high-priority fix at all.

**Comment by ltseng** [comment]
I think @CourtneyLee333 has encountered an issue with seeing screens sometimes taking up to 5 or 10 seconds to respond though, so I'll try to get more details or a screenrecording from her whenever she is able to consistently replicate it.

**Comment by shrouxm** [comment]
that sounds like what i used to experience that originally prompted me to file this issue, but i stopped experiencing it at some point, so would be good to know if it's still happening on other devices!

**Comment by CourtneyLee333** [comment]
It happened on my ipad today, but my app was out of date. I'll keep an eye out for it.

**Comment by ltseng** [comment]
Still replicable on build 46 on iPhone and iPad, not happening on Android

**Comment by ltseng** [comment]
I'm seeing intermittent problems with the map and selecting locations as well on iPhone build 48

https://github.com/techmatters/terraso-mobile-client/assets/2730009/2dd83d53-e1a5-4efa-a7bb-7825b8dd2ad1

**Comment by ltseng** [comment]
This is steadily getting worse for me in testing on both the iPad and iPhone. I'm bumping this up to a priority 2, but we definitely need to resolve this before our first actual release to users @paulschreiber

**Comment by shrouxm** [comment]
i experience this on my testing device as well (iphone 8). it is noticeably worse when battery saver mode is on, which leads me to suspect it's a performance problem, as silly as that seems given how simple the app so far is and how powerful ios devices are. in the past when screens have been slow and i've profiled the issue has been NativeBase, so if that is true of this issue as well i'm not sure what the right way to proceed would be but it makes me think we do need to do _something_ to mitigate the NativeBase situation before release

**Comment by ltseng** [comment]
Thanks for the analysis! I checked my iPhone and am pretty sure Battery Saver is not enabled, but the performance is terrible regardless. I agree that this is a high-priority issue to solve before release though.

**Comment by CourtneyLee333** [comment]
Would that mitigation of NativeBase situation mean switching to a different component library?

**Comment by shrouxm** [comment]
@CourtneyLee333 unclear before looking into it

**Comment by shrouxm** [comment]
going to write what i'm doing as i do it: 

took a [profiling sample](https://drive.google.com/file/d/12oyNVMqZ27VDwFCzRCOVAUOW4xOcVqDk/view?usp=drive_link) according to [these instructions](https://reactnative.dev/docs/profile-hermes). the instructions also indicate how you can load the sample in chrome devtools to look yourself

i took the first profile this way because it was easiest, but the build is in dev mode so that introduces some inaccuracy. when i restrict the timeline to areas of high render activity, it seems pretty clear that nativebase (in particular, the `usePropsResolution` hook which is called on each render by each of their components) is the bottleneck

i'm tempted to just move on to thinking about solutions, but i'm going to take a profile using a different method with the app in a production build just to be thorough

**Comment by shrouxm** [comment]
update: 

i am blocked on doing a production build profile so i started researching solutions

first thing i checked was whether the gluestack/nativebase folks had created an easy migration path from nativebase to gluestack, since they've deprecated nativebase in favor of gluestack

i found [this migration page](https://docs.nativebase.io/migration-guide-gluestack-ui) in the docs which links to [this guide](https://geekyants.notion.site/NativeBase-to-gluestack-ui-Migration-Plan-WIP-6409006d9ecc4fb2a559cca6117a3e7c) which is labeled "WIP"

there is a manual migration method and an auto migration method. i followed the (very sparse) instructions for the auto migration and it is completely broken. the manual migration is what it says on the tin: manually migrating each usage of a nativebase component to the equivalent usage of a gluestack component

at this point i would rather do a manual migration to bare react native than to a new component library from this team, but either way it is a lot of work. i will do some more thinking on if there's a better path forward

**Comment by shrouxm** [comment]
update:

looking around the code a bit more, i am curious to check if it's a compound problem of nativebase components rendering slowly, _and_ we're re-rendering the components too often. i will try to do some profiling/experiments in this vein

**Comment by CourtneyLee333** [comment]
Quick comment on your last item in the list above - I would say we can get rid of the tiny thumbnail if needed. I don't know how much good it's doing for users anyway. The objective was to eventually have a "featured image" the user could upload for the site, with the tiny map image being the default featured image.

**Comment by CourtneyLee333** [comment]
@shrouxm Yes, that will be good.

**Comment by ltseng** [comment]
Question: how much more time is CallStack going to spend on this issue, since there is a draft PR open? 

Steve says that they think they have made a substantial impact with the new PR, and that they recommend it be merged. Waiting for Igor from CallStack to respond with a yes/no answer on whether they are done. There were also no profiling results posted in the PR.

**Comment by igorbej** [comment]
@ltseng @CourtneyLee333 @shrouxm Hi everyone 👋, I'm still looking into a few things related to performance, I plan to post an update in the PR with my findings and some profiling results this week (perhaps tomorrow), hope that's okay with y'all, cheers 😄

**Comment by CourtneyLee333** [comment]
Hi 👋, and thank you!

**Comment by igorbej** [comment]
@CourtneyLee333 @ltseng @shrouxm Hey folks, I updated the description of https://github.com/techmatters/terraso-mobile-client/pull/787, I plan to add a couple more things next week depending on my availability, but I hope what I've written so far will be useful to you, cheers!

**Comment by ltseng** [comment]
Awesome, thank you! Wonderful news and writeup. I appreciate all the time you've spent on this, and the actionable insights you have provided.
